### ISSUE TITLE
[BUG] CSRs inaccessible: hpmcountern 

### ISSUE THREAD

**Issue by flaviens** [openhwgroup/cva6#1213]
### Is there an existing CVA6 bug for this?

- [X] I have searched the existing bug issues

### Bug Description

Hi there!

The riscv privileged specification says:

> Note that although CSRs and instructions are associated with one privilege level, they are also accessible at all higher privilege levels.

[The CVA6 CSR specification](https://github.com/openhwgroup/cva6/blob/master/docs/01_cva6_user/ip-xact/cva6_csr.md) says about `scounteren`:
> When HPMn is clear, attempts to read the hpmcountern register while executing in U-mode will cause an illegal instruction exception. When this bit is set, access to the corresponding register is permitted.

Hence, I understand that setting the correct bits of `scountern` should provide read access to `hpmcountern` in user mode, and hence in particular in machine mode.

### Example snippet

This snippet should store 1 to address 0x10 but stores 0 instead.
I checked the behavior with Spike, which does not throw an exception on reading hpmcounter3.
In contrast, CVA6 throws an illegal instruction exception.

```
  .section ".text.init","ax",@progbits
  .globl _start
  .align 2
_start:

  la 	  ra, trap_handler
  csrrw	zero,mtvec,ra

  addi  t6, zero, -1
  csrw  scounteren, t6
  csrr   zero,hpmcounter3

  li a0, 0x10
  li t0, 0

  sd t0, (a0)
  sd zero, (zero)

infinite_loop0:
  j infinite_loop0

.section ".trap_handler","ax",@progbits
trap_handler:

  li a0, 0x10
  li t0, 1

  sd t0, (a0)
  sd zero, (zero)

infinite_loop1:
  j infinite_loop1
```

Thanks!

**Comment by MikeOpenHWGroup** [comment]
Hi @spidugu444.  I have assigned this issue to you since you are the Lead on CSR verification for CV32A60X.  I know we are not verifying the performance counters for v5 of that core but this is an interesting find that perhaps we should reproduce.

**Comment by spidugu444** [comment]
Hi @flaviens, could you please provide the details to reproduce this issue.

**Comment by flaviens** [comment]
Hi @spidugu444 , what details do you need?

**Comment by spidugu444** [comment]
@flaviens, Details on how to reproduce this issue

**Comment by flaviens** [comment]
What do you need more than what I provided that would help you run the issue?
1. Are you able to run the code on CVA6?
2. Are you having any trouble to reproduce the behavior?

**Comment by JeanRochCoulon** [comment]
@TulikaSi   Can you investigate to know whether this bug is currently in the RTL ?

**Comment by TulikaSi** [comment]
Hi @flaviens @JeanRochCoulon 
The hpmcounters have been implemented in CVA6.
@flaviens Can you try to run the test case once again and check if it is throwing the error still.

**Comment by flaviens** [comment]
Can you please try to use Cascade and see how it goes? Please let me know if you encounter any difficulty.

**Comment by TulikaSi** [comment]
Hi @flaviens  
We ran a small testcase to read the **hpmcounter3** in user mode by enabling the 4th bit of mcountern and scounteren CSR. hpmcounter is a shadow register that just gives the value what is being read by mhpmcounter for the user mode. I have attached the snips of the logfile  for your reference. As we can see from the log file that the current RTL behaves properly and does not throw illegal instruction exception.
Also, the PR (https://github.com/openhwgroup/cva6/pull/1227 ) has been already done for the implementation of the shadow registers.

![hpmcounter_log](https://github.com/openhwgroup/cva6/assets/98868403/46ba2189-444e-4e58-8869-2ffc20e99257)

I hope it helps and we can close this issue.

@JeanRochCoulon
